docs/madr/decisions/057-rework-mesh-routes-matching.md

```diff
@@ -0,0 +1,741 @@
+spec:
+  targetRef:
+    kind: MeshHTTPRoute
+    name: slow-backend-route
+    namespace: backend-ns
+  default:
+    http:
+      requestTimeout: 15s
+```
+
+## Mesh*Route in spec.to[].targetRef design
+
+The design is grounded in multiple recent MADRs around improvements to policies: 
+
+- [MeshService policy matching](https://github.com/kumahq/kuma/blob/master/docs/madr/decisions/046-meshservice-policy-matching.md)
+- [Policy on Namespace](https://github.com/kumahq/kuma/blob/master/docs/madr/decisions/052-policy-on-namespace.md) - which introduced consumer/producer policies
+- [Policy matching with real resources](https://github.com/kumahq/kuma/pull/10726/files)
+
+### API
+
+New API for policy that selects `Mesh*Route`
+
+```yaml
+type: MeshTimeout
+name: timeouts-on-route
+mesh: default
+spec:
+  targetRef:
+    kind: Mesh | MeshSubset # either Mesh, MeshSubset or MeshGateway
+  to:
+    - targetRef:
+        kind: MeshHTTPRoute | MeshTCPRoute # either MeshHTTPRoute or MeshTCPRoute
+        name: route-to-backend
+      default:
+        http:
+          requestTimeout: 5s
+          streamIdleTimeout: 1h
+```
+
+specifying `Mesh` in `spec.targetRef` will be optional (this probably will be most used case). Minimal policy, targeting
+`MeshHTTPRoute` will look like this:
+
+```yaml
+type: MeshTimeout
+name: timeouts-on-route
+mesh: default
+spec:
+  to:
+    - targetRef:
+        kind: MeshHTTPRoute
+        name: route-to-backend
+      default:
+        http:
+          requestTimeout: 5s
+          streamIdleTimeout: 1h
+```
+
+To introduce this change we need to deprecate specifying `Mesh*Route` in `spec.targetRef`.
+
+Should we allow selecting MeshGateway in `spec.targetRef` when `Mesh*Route` is configured in `spec.to[].targetRef`? probably this 
+is obsolete as Route will select gateway, and we want to apply policy on a route. Is there a need to make it more specific than route? 
+
+### Affected policies
+
+At the moment we can only target `Mesh*Routes` in MeshTimeout policy. This MADR is based on MeshTimeout policy.
+Looking at what can be configured at Envoy route we can also apply this to our MeshRetry policy. Rate limit is also configurable on route,
+but we are configuring rate limit on inbound traffic, which does not apply to this MADR. 
+
+#### Producer/consumer model for targeting Mesh*Route in other policies
+
+It is worth to mention that rate limit would not work with our producer/consumer model. In our model consumer always has more priority
+but consumer should not override producer rate limit as owner of the service should know best, the amount of traffic it could handle.
+
+For MeshRetry our producer/consumer makes sense since producer config is a suggestion from owner of the service. If consumer
+wants to override it because they don't want to retry, they just want to fail fast or the want to retry on bigger number of errors
+they can configure it by applying consumer policy.
+
+### Merging and applying configurations
```

Comment:
What about the case when MeshHTTPRoute has multiple to-items to different MSs?
```yaml
kind: MeshHTTPRoute
metadata:
  name: route-1
spec:
  to:
    - targetRef:
        kind: MeshService
        name: backend
      rules: [...]
    - targetRef:
        kind: MeshService
        name: frontend
      rules: [...]
```

Can I reference `route-1` from the policy in this case? Does it apply to both outbounds?

Reply:
So this is problematic, I think that in perfect world we should merge confs for selected resource since different resources correspond with different Envoy config. So I think when merging we should not treat Mesh
*Route as subset of MeshService and merge conf with MeshService so for policy:

```yaml
apiVersion: kuma.io/v1alpha1
kind: MeshTimeout
metadata:
  name: timeout-on-backend
  namespace: backend-ns
  labels:
    kuma.io/mesh: default
spec:
  to:
  - targetRef:
      kind: MeshService
      name: backend
    default:
      http:
        streamIdleTimeout: 1h
        requestTimeout: 5s
   - targetRef:
        kind: MeshService
        name: frontend
     default:
       http:
         streamIdleTimeout: 2h
         requestTimeout: 10s
  - targetRef:
      kind: MeshHTTPRoute
      name: slow-backend-route
      namespace: backend-ns
    default:
      http:
        requestTimeout: 15s
```

after merging, we should have:

```yaml
resourceRules:
  backend.backend-ns:
    conf:
      http:
        streamIdleTimeout: 1h
        requestTimeout: 5s
  frontend.backend-ns:
    conf:
      http:
        streamIdleTimeout: 2h
        requestTimeout: 10s
  slow-backend-route.backend-ns:
    conf:
      http:
        requestTimeout: 15s
```

And when applying configuration, we should take this into account. So we will have two outbounds for each MeshService, and they will share a route. So take all of this information into account, and combine config for MeshService with Mesh*Route for given outbound pointing to specific service

Reply:
I've updated mard as discussed